File: backend/app/schemas/reminder.py
from datetime import date, datetime
from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


ReminderKind = Literal[
    "mot",
    "maintenance",
]


ReminderSeverity = Literal[
    "urgent",
    "warning",
]


class ReminderRead(BaseModel):
    reminder_key: str

    kind: ReminderKind
    severity: ReminderSeverity

    title: str
    message: str

    vehicle_id: int

    registration: str

    make: str | None
    model: str | None

    due_date: date | None = None
    due_mileage: int | None = None
    current_mileage: int | None = None

    action_href: str


class ReminderSummaryRead(BaseModel):
    total: int

    urgent: int
    warning: int

    mot: int
    maintenance: int


class ReminderSettingsRead(BaseModel):
    user_id: int

    mot_enabled: bool
    maintenance_enabled: bool

    due_soon_days: int
    due_soon_miles: int

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class ReminderSettingsUpdate(BaseModel):
    mot_enabled: bool | None = None

    maintenance_enabled: bool | None = None

    due_soon_days: int | None = Field(
        default=None,
        ge=1,
        le=90,
    )

    due_soon_miles: int | None = Field(
        default=None,
        ge=100,
        le=5000,
    )


class ReminderDismissRequest(BaseModel):
    reminder_key: str = Field(
        min_length=1,
        max_length=255,
    )

File: frontend/src/types/reminder.ts
export type ReminderKind =
  | "mot"
  | "maintenance";


export type ReminderSeverity =
  | "urgent"
  | "warning";


export interface Reminder {
  reminder_key: string;

  kind: ReminderKind;
  severity: ReminderSeverity;

  title: string;
  message: string;

  vehicle_id: number;

  registration: string;

  make: string | null;
  model: string | null;

  due_date: string | null;

  due_mileage: number | null;
  current_mileage: number | null;

  action_href: string;
}


export interface ReminderSummary {
  total: number;

  urgent: number;
  warning: number;

  mot: number;
  maintenance: number;
}


export interface ReminderSettings {
  user_id: number;

  mot_enabled: boolean;
  maintenance_enabled: boolean;

  due_soon_days: number;
  due_soon_miles: number;

  created_at: string;
  updated_at: string;
}


export interface ReminderSettingsPayload {
  mot_enabled?: boolean;

  maintenance_enabled?: boolean;

  due_soon_days?: number;
  due_soon_miles?: number;
}